grakn-factory/titan-factory/src/main/java/ai/grakn/factory/TitanPreviousPropertyStepStrategy.java

```diff
@@ -0,0 +1,121 @@
+
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * Optimisation applied to use Titan indices in the following additional case:
+ * <p>
+ * <code>
+ * g.V().outE().values(c).as(b).V().filter(__.properties(a).where(P.eq(b)));
+ * </code>
+ * <p>
+ * In this instance, the vertex can be looked up directly in Titan, joining the {@code V().filter(..)}
+ * steps together.
+ *
+ * @author Felix Chapman
+ */
+public class TitanPreviousPropertyStepStrategy
+        extends AbstractTraversalStrategy<ProviderOptimizationStrategy> implements ProviderOptimizationStrategy {
+
+    private static final long serialVersionUID = 6888929702831948298L;
+
+    @Override
+    public void apply(Traversal.Admin<?, ?> traversal) {
+
+        // Retrieve all graph (`V()`) steps - this is the step the strategy should replace
+        List<TitanGraphStep> graphSteps = TraversalHelper.getStepsOfClass(TitanGraphStep.class, traversal);
+
+        for (TitanGraphStep graphStep : graphSteps) {
+            // For each graph step, confirm it follows this pattern:
+            // `V().filter(__.properties(a).where(P.eq(b)))`
+
+            if (!(graphStep.getNextStep() instanceof TraversalFilterStep)) continue;
+            TraversalFilterStep<Vertex> filterStep = (TraversalFilterStep<Vertex>) graphStep.getNextStep();
+
+            // Retrieve the filter steps e.g. `__.properties(a).where(P.eq(b))`
+            List<Step> steps = stepsFromFilterStep(filterStep);
+
+            if (steps.size() < 2) continue;
+            Step propertiesStep = steps.get(0); // This is `properties(a)`
+            Step whereStep = steps.get(1);      // This is `filter(__.where(P.eq(b)))`
+
+            // Get the property key `a`
+            if (!(propertiesStep instanceof PropertiesStep)) continue;
+            Optional<String> propertyKey = propertyFromPropertiesStep((PropertiesStep<Vertex>) propertiesStep);
+            if (!propertyKey.isPresent()) continue;
+
+            // Get the step label `b`
+            if (!(whereStep instanceof WherePredicateStep)) continue;
+            Optional<String> label = eqPredicateFromWherePredicate((WherePredicateStep<Vertex>) whereStep);
+            if (!label.isPresent()) continue;
+
+            executeStrategy(traversal, graphStep, filterStep, propertyKey.get(), label.get());
+        }
+    }
+
+    private List<Step> stepsFromFilterStep(TraversalFilterStep<Vertex> filterStep) {
+        // TraversalFilterStep always has exactly one child, so this is safe
+        return filterStep.getLocalChildren().get(0).getSteps();
+    }
+
+    private Optional<String> propertyFromPropertiesStep(PropertiesStep<Vertex> propertiesStep) {
+        String[] propertyKeys = propertiesStep.getPropertyKeys();
+        if (propertyKeys.length != 1) return Optional.empty();
+        return Optional.of(propertyKeys[0]);
+    }
+
+    private Optional<String> eqPredicateFromWherePredicate(WherePredicateStep<Vertex> whereStep) {
```

Comment:
maybe this method can be called `getLabelFromEqualsPredicate` I think that makes more sense.